app/demo-geant-integration/HitRootIO.cc

```diff
@@ -0,0 +1,202 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file demo-geant-integration/HitRootIO.cc
+//---------------------------------------------------------------------------//
+#include "HitRootIO.hh"
+
+#include <regex>
+#include <G4Event.hh>
+#include <G4RunManager.hh>
+#include <G4Threading.hh>
+#include <TBranch.h>
+#include <TFile.h>
+#include <TObject.h>
+#include <TROOT.h>
+#include <TTree.h>
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/io/Logger.hh"
+#include "accel/ExceptionConverter.hh"
+#include "accel/SetupOptions.hh"
+
+#include "GlobalSetup.hh"
+#include "SensitiveHit.hh"
+
+namespace demo_geant
+{
+G4ThreadLocal HitRootIO* HitRootIO::instance_ = nullptr;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create a ROOT output file for each worker except the master thread if MT
+ */
+HitRootIO::HitRootIO()
+{
+    ROOT::EnableThreadSafety();
+
+    file_name_ = std::regex_replace(
+        GlobalSetup::Instance()->GetSetupOptions()->output_file,
+        std::regex("\\.json$"),
+        ".root");
```

Comment:
Isn't it better to just add an extra field in the macro defining the root output name?

Reply:
Actually I do like that idea better...

Reply:
> Isn't it better to just add an extra field in the macro defining the root output name?

This makes sense if we want to support writing both type of files in the same execution *and* want to allow for the same to be completely different.

If not, we could also have the "output_file" be "just" the name without the extension and the writer add the proper extension.  Ie. the code above becomes:
```
file_name_ = GlobalSetup::Instance()->GetSetupOptions()->output_file + ".root";
```

Reply:
I think adding the setting is the right approach: that way if the user specifies a non-empty filename it'll throw an exception instead of the file just not being there.

Reply:
We can defer that addition to another commit.